File: letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        dict = {
            "2" : ["a", "b", "c"],
            "3" : ["d", "e", "f"],
            "4" : ["g", "h", "i"],
            "5" : ["j", "k", "l"],
            "6" : ["m", "n", "o"],
            "7" : ["p", "q", "r", "s"],
            "8" : ["t", "u", "v"],
            "9" : ["w", "x", "y", "z"]
        }

        def dfs(s: str, arr: List) -> List:
            if not s:
                return arr
            if not arr:
                arr = dfs(s[1:], dict[s[0]])
            else:
                arr = dfs(s[1:], list(map(''.join, itertools.product(arr, dict[s[0]]))))
            return arr

        return list(dfs(digits, []))

File: letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (digits.length() == 0) return result;
        HashMap<Character, String> dict = new HashMap<Character, String>() {
            {
                put('2', "abc"); put('3',"def");
                put('4',"ghi"); put('5',"jkl");
                put('6',"mno"); put('7',"pqrs");
                put('8',"tuv"); put('9',"wxyz");
                }
            };
        findCombinations(digits, result, dict, 0, "");
        return result;
    }

    private void findCombinations(String digits, List<String> result,
                                  HashMap<Character, String> dict, int start, String s) {
        if (start == digits.length()) {
            result.add(s);
            return;
        }

        String temp = dict.get(digits.charAt(start));
        for (int j = 0; j < temp.length(); j++) {
            findCombinations(digits, result, dict, start + 1, s + temp.charAt(j));
        }
    }
}